README.md
# CR14 RFID reader driver for Linux (Raspbian)

[![GitHub Super-Linter](https://github.com/pguyot/cr14/actions/workflows/super-linter.yml/badge.svg)](https://github.com/marketplace/actions/super-linter)
[![ARM Runner](https://github.com/pguyot/cr14/actions/workflows/arm-runner.yml/badge.svg)](https://github.com/marketplace/actions/arm-runner)
[![Codacy](https://app.codacy.com/project/badge/Grade/dcee41ebf0a9416182d54051edfe8b5f)](https://www.codacy.com/gh/pguyot/cr14/dashboard)

## Datasheet and technical documents

Chipset:

- [CR14](https://datasheet.octopart.com/CR14-MQP/1GE-STMicroelectronics-datasheet-10836722.pdf)

PICC:

- [SRIX4K](http://www.orangetags.com/wp-content/downloads/datasheet/STM/srix4k.pdf)
- [SRI512](https://www.advanide.de/wp-content/uploads/products/rfid/SRI512.pdf)
- [SRT512](https://www.advanide.de/wp-content/uploads/products/rfid/SRT512.pdf)
- [SRI4K](https://www.advanide.de/wp-content/uploads/products/rfid/SRI4K.pdf)
- [SRI2K](https://www.advanide.de/wp-content/uploads/products/rfid/SRI2K.pdf)
- [ST25TB512-AC](https://www.st.com/resource/en/datasheet/st25tb512-ac.pdf)
- [ST25TB04K](https://www.st.com/resource/en/datasheet/st25tb04k.pdf)
- [ST25TB512-AT](https://www.st.com/resource/en/datasheet/st25tb512-at.pdf)
- [ST25TB02K](https://www.st.com/resource/en/datasheet/st25tb02k.pdf)

This driver may be compatible with CRX14 as well.
Tested with CR14 and SRI512/SRT512 PICCs.

## Installation

Install requirements

    sudo apt-get install raspberrypi-kernel-headers

Clone source code.

    git clone https://github.com/pguyot/cr14

Compile and install with

    cd cr14
    make
    sudo make install

Makefile will automatically edit /boot/config.txt and add/enable if required
the following params and overlays:

    dtparam=i2c_arm=on
    dtoverlay=cr14

You might want to review changes before rebooting.

Reboot.

## Interface

Driver creates device /dev/rfid0

Several modes are available, see the sample Python scripts in examples.

Simplest mode consists in opening device read-only. The CR14 will be polled
every 0.5 second, printing detected tag UIDs preceeded by 'u' (UIDs are printed
in little endian, LSB first).

More complex interactions are possible by opening device r/w and sending
commands, for example to read or write EEPROM.


debian/changelog
cr14 (1.1) unstable; urgency=medium

  * Fixed some minor errors revealed by linters.
  * Driver now detects ST25R

cr14 (1.0) unstable; urgency=medium

  * Initial Release.

 -- Guilhem Lettron <guilhem@barpilot.io>  Sat, 05 Sep 2020 13:01:26 +0200


examples/write_block.py
#!/usr/bin/env python3

import os

# Example code demonstrating how to read and write single block.
# Read block #7 and write it back, as a counter.

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid_le = packet[1:]
        uid = bytearray(uid_le)
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        os.write(rfid, b"r" + uid_le + b"\x07")
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"r":
            data = os.read(rfid, 4)
            counter = int.from_bytes(data, byteorder="little")
            print(f"Counter = {counter}, decrementing it")
            counter = counter - 1
            data = counter.to_bytes(4, byteorder="little", signed=counter < 0)
            os.write(rfid, b"w" + uid_le + b"\x07" + data)
            packet = os.read(rfid, 1)
            if packet == b"w":
                written_data = os.read(rfid, 4)
                if written_data != data:
                    print(
                        f"Data mismatch, got {written_data.hex()}"
                        f" but wrote {data.hex()}"
                    )
            else:
                print(f"Unexpected packet, got {packet.hex()}, expected w")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/dump_counter.py
#!/usr/bin/env python3

import os

# Example code demonstrating how to read multiple blocks.
# Print the counter values at blocks 5 and 6.

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid = bytearray(packet[1:])
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        os.write(rfid, b"R" + packet[1:] + b"\x02\x05\x06")
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"R":
            count = os.read(rfid, 1)
            if count[0] != 2:
                print(f"Unexpected block counts, got {count[0]}, expected 2")
            for x in range(5, 7):
                data = os.read(rfid, 4)
                counter_value = int.from_bytes(data, byteorder="little")
                print(f"{x} counter={counter_value} ({data.hex()})")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/erase_blocks.py
#!/usr/bin/env python3

import os

# Example code demonstrating how to write several blocks in a row.
# Write FFFFFFFF to blocks 7 to 9 (that may be affected by other scripts)

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid_le = packet[1:]
        uid = bytearray(uid_le)
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        erased_data = b"\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF"
        os.write(rfid, b"W" + uid_le + b"\x03\x07\x08\x09" + erased_data)
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"W":
            count = os.read(rfid, 1)
            if count[0] != 3:
                print(f"Unexpected count, got {count[0]}, expected 3")
            else:
                data = os.read(rfid, 12)
                if data != erased_data:
                    print(
                        f"Data mismatch, got {data.hex()} but wrote {erased_data.hex()}"
                    )
                else:
                    print("Erased blocks 7 to 9 (wrote FFFFFFFF)")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/write_blocks.py
#!/usr/bin/env python3

import os

# Example code demonstrating how to read and write several blocks in a row.
# Read blocks #7, #8 and #9 and swap them.

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid_le = packet[1:]
        uid = bytearray(uid_le)
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        os.write(rfid, b"R" + uid_le + b"\x03\x07\x08\x09")
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"R":
            count = os.read(rfid, 1)
            if count[0] != 3:
                print(f"Unexpected count, got {count[0]}, expected 3")
            else:
                block7 = os.read(rfid, 4)
                block8 = os.read(rfid, 4)
                block9 = os.read(rfid, 4)
                data = block7 + block8 + block9
                os.write(
                    rfid,
                    b"W" + uid_le + b"\x03\x07\x08\x09" + data,
                )
                packet = os.read(rfid, 1)
                if packet == b"W":
                    count = os.read(rfid, 1)
                    if count[0] != 3:
                        print(f"Unexpected count, got {count[0]}, expected 3")
                    else:
                        written7 = os.read(rfid, 4)
                        written8 = os.read(rfid, 4)
                        written9 = os.read(rfid, 4)
                        if written7 != block8:
                            print(
                                f"Data mismatch, got {written7.hex()}"
                                f" but wrote {block8.hex()}"
                            )
                        if written8 != block9:
                            print(
                                f"Data mismatch, got {written7.hex()}"
                                f" but wrote {block8.hex()}"
                            )
                        if written9 != block7:
                            print(
                                f"Data mismatch, got {written7.hex()}"
                                f" but wrote {block8.hex()}"
                            )
                        print(f"New 7: {written7.hex()}")
                        print(f"New 8: {written8.hex()}")
                        print(f"New 9: {written9.hex()}")
                else:
                    print(f"Unexpected packet, got {packet.hex()}, expected w")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/dump_512_bytes.py
#!/usr/bin/env python3

import os

# Example code demonstrating how to read multiple blocks.
# Dumps the first 512 bytes (typically every byte for 512 bytes PICCs)

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid = bytearray(packet[1:])
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        os.write(
            rfid,
            b"R"
            + packet[1:]
            + b"\x10\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F",
        )
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"R":
            count = os.read(rfid, 1)
            if count[0] != 16:
                print(f"Unexpected block counts, got {count[0]}, expected 16")
            for x in range(0, count[0]):
                data = os.read(rfid, 4)
                data = bytearray(data)
                data.reverse()
                data_str = ":".join("{:02x}".format(c) for c in data)
                print(f"{x} {data_str}")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/read_system_block.py
#!/usr/bin/env python3

import os

rfid = os.open("/dev/rfid0", os.O_RDWR)
print("Waiting for a chip")
try:
    os.write(rfid, b"p")
    packet = os.read(rfid, 9)
    if packet[0] != ord("u"):
        print(f"Unexpected packet header {packet[0]}")
    else:
        # uid is in little endian
        uid_le = packet[1:]
        uid_be = bytearray(uid_le)
        uid_be.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid_be)
        print(f"UID: {uid_str}")
        if uid_be[0] != 0xD0:
            print(f"Unexpected MSB, got {uid_be[0]}")
        if uid_be[1] != 0x02:
            print("Not a STMicroelectronics chip, will read block 255 anyway")
        os.write(rfid, b"r" + uid_le + b"\xFF")
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"r":
            data = os.read(rfid, 4)
            data = bytearray(data)
            data.reverse()
            data_str = ":".join("{:02x}".format(c) for c in data)
            print(f"Data (read single): {data_str}")
        else:
            print(f"Unexpected packet header {packet[0]}")
        os.write(rfid, b"R" + uid_le + b"\x01\xFF")
        packet = os.read(rfid, 1)
        while packet == b"u":
            os.read(rfid, 8)
            packet = os.read(rfid, 1)
        if packet == b"R":
            count = os.read(rfid, 1)
            if count[0] != 1:
                print(f"Unexpected count, got {count[0]}, expected 1")
            data = os.read(rfid, 4)
            data = bytearray(data)
            data.reverse()
            data_str = ":".join("{:02x}".format(c) for c in data)
            print(f"Data (read block): {data_str}")
        else:
            print(f"Unexpected packet header {packet[0]}")
except KeyboardInterrupt:
    pass
os.close(rfid)


examples/simple_uid_reader.py
#!/usr/bin/env python3

import os

MODELS = {
    0x02: {
        (
            6,
            0b000011,
            "SRIX4K",
        ),  # http://www.orangetags.com/wp-content/downloads/datasheet/STM/srix4k.pdf
        (
            6,
            0b000110,
            "SRI512",
        ),  # http://www.advanide.com/wp-content/uploads/products/rfid/SRI512.pdf
        (
            6,
            0b001100,
            "SRT512",
        ),  # https://www.advanide.de/wp-content/uploads/products/rfid/SRT512.pdf
        (
            6,
            0b000111,
            "SRI4K",
        ),  # https://www.advanide.de/wp-content/uploads/products/rfid/SRI4K.pdf
        (
            6,
            0b001111,
            "SRI2K",
        ),  # https://www.advanide.de/wp-content/uploads/products/rfid/SRI2K.pdf
        (
            8,
            0x1B,
            "ST25TB512-AC",
        ),  # https://www.st.com/resource/en/datasheet/st25tb512-ac.pdf
        (
            8,
            0x1F,
            "ST25TB04K",
        ),  # https://www.st.com/resource/en/datasheet/st25tb04k.pdf
        (
            8,
            0x33,
            "ST25TB512-AT",
        ),  # https://www.st.com/resource/en/datasheet/st25tb512-at.pdf
        (
            8,
            0x3F,
            "ST25TB02K",
        ),  # https://www.st.com/resource/en/datasheet/st25tb02k.pdf
    }
}
MANUFACTURERS = {
    0x01: "Motorola",
    0x02: "ST Microelectronics",
    0x03: "Hitachi",
    0x04: "NXP Semiconductors",
    0x05: "Infineon Technologies",
    0x06: "Cylinc",
    0x07: "Texas Instruments Tag-it",
    0x08: "Fujitsu Limited",
    0x09: "Matsushita Electric Industrial",
    0x0A: "NEC",
    0x0B: "Oki Electric",
    0x0C: "Toshiba",
    0x0D: "Mitsubishi Electric",
    0x0E: "Samsung Electronics",
    0x0F: "Hyundai Electronics",
    0x10: "LG Semiconductors",
    0x16: "EM Microelectronic-Marin",
    0x1F: "Melexis",
    0x2B: "Maxim",
    0x33: "AMIC",
    0x44: "GenTag, Inc (USA)",
    0x45: "Invengo Information Technology Co.Ltd",
}

rfid = os.open("/dev/rfid0", os.O_RDONLY)
print("Exit with control-C")
while True:
    try:
        packet = os.read(rfid, 9)
        if packet[0] != ord("u"):
            print(f"Unexpected packet header {packet[0]}")
            break
        # uid is in little endian
        uid = bytearray(packet[1:])
        uid.reverse()
        uid_str = ":".join("{:02x}".format(c) for c in uid)
        print(f"UID: {uid_str}")
        if uid[0] != 0xD0:
            print(f"Unexpected MSB, got {uid[0]}")
        if uid[1] in MANUFACTURERS:
            manufacturer = MANUFACTURERS[uid[1]]
            print(f"Manufacturer: {manufacturer}")
        else:
            print(f"Manufacturer: unknown ({int(uid[1])})")
        serial_start = 2
        if uid[1] in MODELS:
            model_unknown = True
            for bits, model_id, model_str in MODELS[uid[1]]:
                model = uid[2]
                if bits < 8:
                    model = model >> (8 - bits)
                if model == model_id:
                    print(f"Model: {model_str}")
                    if bits < 8:
                        uid[2] = uid[2] & ~(model << (8 - bits))
                    else:
                        serial_start = 3
                    model_unknown = False
                    break
            if model_unknown:
                print(f"Model: unknown ({uid[2]})")
        serial = ":".join("{:02x}".format(c) for c in uid[serial_start:])
        print(f"Serial number: {serial}")
    except KeyboardInterrupt:
        break
os.close(rfid)
